paho/message_ids.go

```diff
@@ -41,20 +41,29 @@ type CPContext struct {
 // to messages with a messageid
 type MIDs struct {
 	sync.Mutex
-	index map[uint16]*CPContext
+	index      map[uint16]*CPContext
+	lastIdUsed uint16
 }
 
 // Request is the library provided MIDService's implementation of
 // the required interface function()
 func (m *MIDs) Request(c *CPContext) (uint16, error) {
 	m.Lock()
 	defer m.Unlock()
+
+	if m.lastIdUsed < 65535 {
```

Comment:
You must store the c in the map.
And you can't do this if, you need always check the map.

```go
for i, n := c.lastID, 0; n < 65535; i, n = i+1, n+1 {
   if i == 0 {
       i = 1
    }
    if _, ok := ...; ok {
        m.index[i] = c
        m.lastID = i
        return i, nil
    }
}
return 0, ErrNoMoreIDs
```

Reply:
You were probably reading just the first commit, otherwise you wouldn't have written the stuff about 'need to check blah blah'. Very nice loop! Love the `uint16` overflow! :)